streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -1097,10 +1102,10 @@ void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
             for (final Task task : commitNeededActiveTasks) {
                 if (!dirtyTasks.contains(task)) {
                     try {
-                        // for non-revoking active tasks, we should not enforce checkpoint
-                        // since if it is EOS enabled, no checkpoint should be written while
-                        // the task is in RUNNING tate
-                        task.postCommit(false);
+                        // we only enforce a checkpoint if the transaction buffers are full
+                        // to avoid unnecessary flushing of stores under EOS
+                        final boolean enforceCheckpoint = maxUncommittedStateBytes > -1 && tasks.approximateUncommittedStateBytes() >= maxUncommittedStateBytes;
+                        task.postCommit(enforceCheckpoint);
```

Comment:
Also here, the tasks should have been committed. Why are transaction buffers not empty?

Reply:
Same as above.

That said, I'm not convinced this particular change is necessary. This code is run during rebalance, and AFAIK no records will be processing during a rebalance; it shouldn't be possible for the transaction buffer to have increased beyond the last time we checked in the `StreamThread` main run-loop (aka. `StreamThread#maybeCommit`).

Worst case scenario, if we leave this as `task.postCommit(false)`, and the state stores aren't flushed during a rebalance, if the transaction buffer is already close to exceeding its capacity, it will be detected during the next iteration of the run-loop.